### PR TITLE
fix(mac): move keyboard menu items to main Input Menu from submenu 🔥

### LINKED ISSUE #9684: bug(mac): Unable to switch keyboards with macOS Sonoma 14.0
### Describe the bug

I encountered difficulty switching between keyboards within the Keyman menu.

### Reproduce the bug

1. Install Keyman 17.0.185 alpha build in macOS Sonoma 14.0.
2. Download and install two keyboards like 'KhmerAngkor' and 'Telugu Basic' using the Keyman Configuration dialog. 
3. Open the Keyman menu from the taskbar. 
4. Verify that there are two keyboards 'KhmerAngkor' and the 'Telugu Basic' Keyboard appears from the Keyboards list. The Khmer Angkor keyboard is selected by default. 
5. Switch to the 'Telugu Basic' keyboard by clicking on the 'Telugu Basic' name. 

Here, I noticed that the keyboard does not switch to 'Telugu Basic'. Seems to be an issue. 

Note: I was able to reproduce it in Keyman 16.0141 Stable build. 

   

### Expected behavior

Switching between any two keyboards should be working fine using Keyman menu list. 

### Related issues

#8599

### Keyman apps

- [ ] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [X] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

16.0.141 Stable build, 17.0.185-alpha build

### Operating system

macOS 14 Sonoma

### Device

_No response_

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### LINKED ISSUE #9864: feat(web): Make popups and keytips children of .kmw-keyboard to streamline CSS selection
### Is your feature request related to a problem? Please describe.

Today, I'm playing with CSS selectors for popups in α17. I've wanted my vowel popups to be colored like my static vowel keys, and I hear from @mcdurdin that we have some new popup styling options in 17.

image

```
.android .kmw-keyboard-sil_cameroon_qwerty .kmw-key[id$='K_A']:not(.kmw-key-touched),
.android #kmw-popup-keys .kmw-key[id*='U_2C6D']:not(.kmw-key-touched) {background-image: radial-gradient(transparent, #00000030), linear-gradient(#7f7f7f55, #beab40dd);}
```
I got it working in mobile preview (haven't tested it on real phones), but I did run into one inconsistency related to selectors.

I usually include the keyboard name in my CSS selector to avoid leakage and give my rules a higher specification priority:
```
.android .kmw-keyboard-sil_cameroon_qwerty .kmw-key[id$='K_A']:not(.kmw-key-touched)
```
but if I add `.kmw-keyboard-sil_cameroon_qwerty` to my popup, it doesn't work. 
```
/* works */
.android #kmw-popup-keys .kmw-key[id*='U_2C6D']:not(.kmw-key-touched) {background-image: radial-gradient(transparent, #00000030), linear-gradient(#7f7f7f55, #beab40dd);}
/* does not work */
.android .kmw-keyboard-sil_cameroon_qwerty #kmw-popup-keys .kmw-key[id*='U_2C6D']:not(.kmw-key-touched) {background-image: radial-gradient(transparent, #00000030), linear-gradient(#7f7f7f55, #beab40dd);}
```

This is because popups are children of `#kmw-popup-keys` which is itself a sibling and not  a child of `.kmw-keyboard-sil_cameroon_qwerty`.  Popups and keytips are both siblings of the keyboard `div`.

image

### Describe the solution you'd like

Since there are now several siblings of  `.kmw-keyboard-sil_cameroon_qwerty` that are specific to the selected keyboard, maybe we need a new `div` wrapper around them that can be the new `.kmw-keyboard-sil_cameroon_qwerty`.

### Describe alternatives you've considered

Add a customized class to the siblings, like: `.kmw-popups-sil_cameroon_qwerty` or  `.kmw-keytips-sil_cameroon_qwerty`, but this seems less ideal.

### Related issues

_No response_

### Keyman apps

- [X] Keyman for Android
- [X] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.198

### Operating system

Windows 11

### Device

_No response_

### Target application

_No response_

### Browser

_No response_

### Keyboard name

sil_cameroon_qwerty

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sgschantz**
Previously, Keyman for Mac supported switching of keyboards using a submenu from the Keyman section of the Input Menu:

image

This  change eliminates the submenu and lists the keyboards at the top of the Keyman section of the Input menu:

image

It is helpful to make this change now because the Keyboards submenu is not functioning properly with macOS Sonoma, 14.0 (see #9684).

Fixes #9684

# User Testing

- **TEST_KEYBOARD_SELECTION**: 
Run Keyman for Mac and verify that a single keyboard is selected in the Keyman section of the Input Menu. Also, confirm that the selected keyboard is actually the one that is processing keystrokes when typing in a text document.

- **TEST_CHANGING_KEYBOARDS**: 
Select a different keyboard from the list in the Keyman section of the Input Menu and verify that:
1) The keyboard that was selected now has the checkmark next to it when you open the Input Menu.
2) The previously selected keyboard does not have a checkmark next to it.
3) When typing in a text document, confirm the selected keyboard is the one that is processing keystrokes.

- **TEST_CONFIGURATION_MENU**: 
Select the 'Configuration...' menu item and confirm that it brings up the Configuration window.

- **TEST_SHOW_OSK_MENU**: 
Select the 'On-Screen Keyboard' menu item and confirm that it brings up the On-screen Keyboard window.

- **TEST_ABOUT_MENU**: 
Select the 'About' menu item and confirm that it brings up the About dialog.

- **TEST_ADD_KEYBOARD**: 
Go to the Configuration window and either install a new keyboard or enable one that is already installed by clicking on the checkbox in the leftmost column of the Keyboards pane. Confirm that the newly added keyboard is now listed in the Keyman section of the Input Menu.

- **TEST_REMOVE_SELECTED_KEYBOARD**: 
Go to the Configuration window and disable the keyboard that is currently active in Keyman by unchecking the corresponding checkbox in the leftmost column of the Keyboards pane. Confirm that:
1) the removed keyboard is no longer listed in the Keyman section of the Input Menu
2) that some other keyboard is now selected instead
3) that the newly selected keyboard is being used for key processing when typing in a text document

- **TEST_REMOVE_ALL_KEYBOARDS**: 
Go to the Configuration window and disable all the keyboards that are currently active in Keyman by unchecking the checkboxes in the leftmost column of the Keyboards pane. Go look at the Input Menu and confirm that the only keyboard listed is a placeholder that says: '(No Keyboard Configured)':

image

- **TEST_MENU_PERFORMANCE**: 
When the Input Menu is opened, confirm that the contents of the menu are fully populated and you do not need to open and close the Input Menu a few times for the Keyman section of the menu to appear.

### CODE REVIEW

File: mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodAppDelegate.m [hunk 1]
```diff
@@ -760,90 +760,174 @@ - (void)setContextBuffer:(NSMutableString *)contextBuffer {
...
+
+- (void)removeDynamicKeyboardMenuItems {
+  int numberToRemove = (int) self.menu.numberOfItems - DEFAULT_KEYMAN_MENU_ITEM_COUNT;
+  
+  if (self.debugMode) {
+    NSLog(@"*** removeDynamicKeyboardMenuItems, self.menu.numberOfItems = %ld, number of items to remove = %d", (long)self.menu.numberOfItems, numberToRemove);
+  }
+  
+  if (numberToRemove > 0) {
+    for (int i = 1; i <= numberToRemove; i++) {
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodAppDelegate.m:807]: Can we use a zero-based loop? It's more consistent with all the rest of our loops in our code

```suggestion
    for (int i = 0; i < numberToRemove; i++) {
```

**Review by mcdurdin — APPROVED** [review]
LGTM

This is a great cleanup of the existing code as well as a much needed improvement to the UI. As a bonus, hopefully it means Keyman works on Sonoma as well 😉

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_KEYBOARD_SELECTION (PASSED):** Tested with the attached PR build (Keyman 17.0.192-alpha-test-9777) on Mac Sonoma OS and here is my observation: 1. Installed a keyman keyboard and verified that the single keyboard is selected in the keyman section of the Input menu. 2. I opened a text document and could type with the selected keyman keyboard without having any issues. 


* **TEST_CHANGING_KEYBOARDS (PASSED):** 1. Verified that the selected keyboard has the checkmark next to it when I opened the Input menu. 2. Also, confirmed that the previously selected keyboard did not have a checkmark next to it. 3. Verified that the selected keyboard is the one that is processing keystrokes When typing in a text document.

   


* **TEST_CONFIGURATION_MENU (PASSED):** 1. Confirmed that clicking the 'Configuration' option from the menu item opens the Configuration window. 


* **TEST_SHOW_OSK_MENU (PASSED):** Verified that by selecting 'On-Screen Keyboard' menu item, open up the On-screen keyboard window. 


* **TEST_ABOUT_MENU (PASSED):** Verified that clicking the 'About' menu item, opens the About dialog. 

   


* **TEST_ADD_KEYBOARD (PASSED):** Tested as per the instructions and confirmed that the newly added keyboard is now listed in the Keyman section of the Input Menu.


* **TEST_REMOVE_SELECTED_KEYBOARD (PASSED):** 1. Verified that the removed keyboard is no longer listed in the Keyman section of the Input menu. 2. Instead of that some other keyboard is selected in the menu. 3. Able to type with the newly selected keyboard in a text document. 

   


* **TEST_REMOVE_ALL_KEYBOARDS (PASSED):** 1. Verified that the 'No Keyboard configured' message appears in gray color in the keyman list menu.

  
   


* **TEST_MENU_PERFORMANCE (PASSED):** 1. Verified that the contents of the menu are fully populated and no need to open and close the Input menu a few times for the keyman section of the menu to appear.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.196-alpha](https://keyman.com/downloads/releases/alpha/17.0.196)

**Pull request merged**